rs/utils/validate_eq_derive/src/lib.rs

```diff
@@ -0,0 +1,150 @@
+extern crate proc_macro;
+extern crate quote;
+extern crate syn;
+
+use proc_macro::TokenStream;
+use quote::quote;
+use syn::Data::Struct;
+
+/// Derive of ValidateEq trait.
+/// Derived implementation compares fields of lhs and rhs and returns the first divergence if any.
+/// A field can have one #[validate_eq(Recursive|Skip)] attribute.
+///   - Recursive calls .validate_eq() and returns path + its error in case of divergence
+///   - Skip ignores the field.
+///   - None (default) compares fields using PartialEq and reports their name in case of
+///     divergence.
+
+enum ValidateEqFieldAttr {
+    /// Compare using .eq() and return field name if diverges.
+    PartialEq,
+    /// Call .validate_eq(); in case of deivergence return the field name and the underlying
+    /// divergence error string.
+    Recursive,
+    /// Ignore for ValidateEq
+    Skip,
+}
```

Comment:
Nit: I'd try to name these variants (and the corresponding attribute fields) for the benefit of the user. And as a potential future user of the attribute, my way of looking at this would be that each field can be either fully compared for equality, partially compared for equality, or ignored.

So I would suggest naming the variants something like `Full` (or `Eq`), `Partial` (or `Custom`, or `Discriminated`, etc.) and `Ignore`.

I understand where `PartialEq` comes from, but in this context it is confusing: the first time I read it, I understood `Recursive` to mean `.eq()` (because that's what `#[derive(Eq, PartialEq)]` does and "partial eq" (my brain ignored the formatting) to mean some fields are included in the comparison, some are excluded (i.e. recursively call `.validate_eq()`).

Reply:
Some `ValidateEq` compare all the fields, to `Full` would be misleading. `Ignore` seems good instead of `Skip`

Reply:
You mean `Partial` would be a misleading name for the current `Recursive`?

Feel free to come up with whatever names you think best reflect what's actually being compared. In my mind it's something like (i) full / regular / default / `Eq`; (ii) partial / custom / `ValidateEq`; and (iii) ignore / excluded.

I find that `Recursive` requires me to stop and think, since `Eq` / `PartialEq` are also recursive by default. So I might need to look into what other variants are available before being able to decide what it does.

As for `PartialEq`, as said, there's also the potential for confusion, because it may refer to the `PartialEq` trait or to "partial equality", which is one way of describing `ValidateEq`. I'd find `Eq` a lot more straightforward (even if we only require that the underlying struct implement `PartialEq`; readability trumps accuracy in my book; and you can always clarify this in the documentation and/or API).

Reply:
Both PartialEq and ValidateEq are 1) Partial in a sense that not all fields are necessarily compared 2) Recursive

Reply:
how about smth like (Ignore | ValidateEq | PartialEq) or (ignore | validate_eq | eq)?

Reply:
Either of those would be fine, thanks.